### PR TITLE
[T-6346][15.0][ADD]sale_order_line_clone: clone sale order lines

### PR DESCRIPTION
**Pull request by RogerSans**
This module allows you the option to clone sale order lines.
T-6346

@ValentinVinagre 
@HaraldPanten 
@luis-ron

### CODE REVIEW

**Review by luis-ron — CHANGES_REQUESTED** [review]
I am not sure if it would be like that, maybe Alberto can contribute, but in the README you indicate:

Go to Settings > Users and Company > Users and edit users how can copy sale order lines. Check the "Can Clone Sale Order Line" option in Extra Rights. Go to Sales > Products > Products and into the sales tab you can find the "Can Clone Sale Order Line" option.

As an addition u indicate: "how can copy sale order lines", it should be "who can copy and sale order lines" i think.

Maybe, you could indicate that the module adds a group that allows to indicate who can make use of the function offered by the module.

I mean, in a small company it is normal to configure users individually, but imagine in a large company... this would be done by groups, not by individual users.

I am waiting for your comments.

Thank you,

**Review by luis-ron — APPROVED** [review]
Functional Review: LGTM 👍🏻 

Please check my last coment and it will be fine :)

**Comment by RogerSans** [comment]
> I am not sure if it would be like that, maybe Alberto can contribute, but in the README you indicate:
> 
> Go to Settings > Users and Company > Users and edit users how can copy sale order lines. Check the "Can Clone Sale Order Line" option in Extra Rights. Go to Sales > Products > Products and into the sales tab you can find the "Can Clone Sale Order Line" option.
> 
> As an addition u indicate: "how can copy sale order lines", it should be "who can copy and sale order lines" i think.
> 
> Maybe, you could indicate that the module adds a group that allows to indicate who can make use of the function offered by the module.
> 
> I mean, in a small company it is normal to configure users individually, but imagine in a large company... this would be done by groups, not by individual users.
> 
> I am waiting for your comments.
> 
> Thank you,

Thank you for your reply. All the user permission options are actually groups. If we look for the name of the option in the Users and Companies > Groups section, the group will appear so we can edit it and add users more quickly. I will add this point to the README if you think it is appropriate.

**Comment by RogerSans** [comment]
@Tisho99 Can you review this please?

File: sale_order_line_clone/models/product_template.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        compute="_compute_service_sale_line",
+        store=True,
+        readonly=False,
+    )
+
+    @api.depends("detailed_type")
+    def _compute_service_sale_line(self):
+        for record in self:
+            if record.detailed_type == "service":
+                record.write({'can_clone_sale_line': False})
```
Comment (Tisho99) [inline sale_order_line_clone/models/product_template.py]: The precommit suggests changing it to record.write({"can_clone_sale_line": False})

File: sale_order_line_clone/models/product_template.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            else:
+                record.write({'can_clone_sale_line': True})
```
Comment (Tisho99) [inline sale_order_line_clone/models/product_template.py]: The precommit suggests changing it to record.write({"can_clone_sale_line": True})

File: sale_order_line_clone/README.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+Installation
+============
+
+# Just install.
+
+Configuration
+=============
+
+# Go to *Settings > Users and Company > Users* and edit users how can
```
Comment (Tisho99) [inline sale_order_line_clone/README.rst]: Please, change this to #. In all the readme (add the .)

This is not properly displayed

File: sale_order_line_clone/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2024 Roger Sans <roger.sans@sygel.es>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Sale Order Line Clone",
+    "version": "15.0.1.0.0",
+    "author": "Sygel, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "Sales",
+    "summary": "Custom partner_name.",
```
Comment (Tisho99) [inline sale_order_line_clone/__manifest__.py]: Have you forgotten the summary?

File: sale_order_line_clone/models/product_template.py (hunk 1 above)
Comment (Tisho99) [reply sale_order_line_clone/models/product_template.py]: Apart from that, the standard in compute is using assignment operations

You can do it in one line

record.can_clone_sale_line = record.detailed_type != "service"

File: sale_order_line_clone/models/sale_order.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        store=True,
+        readonly=False,
+    )
+
+    @api.depends("product_id", "product_id.can_clone_sale_line")
+    def _compute_product_sale_line(self):
+        for record in self:
+            if record.product_id:
+                prod_can_clone_sale_line = record.product_id.can_clone_sale_line
+                record.write({"can_clone_sale_line": prod_can_clone_sale_line})
```
Comment (Tisho99) [inline sale_order_line_clone/models/sale_order.py]: Same as above, why do you use the write?

File: sale_order_line_clone/readme/CONFIGURE.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+\# Go to *Settings \> Users and Company \> Users* and edit users how can
```
Comment (Tisho99) [inline sale_order_line_clone/readme/CONFIGURE.md]: Please, consider replacing "how" with "that"

File: sale_order_line_clone/security/sale_order_line_clone_security.xml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+        <record model="res.groups" id="group_can_clone_sale_line">
+            <field name="name">Can Clone Sale Order Line</field>
+            <field name="category_id" ref="base.module_category_usability" />
+            <field
+            name="users"
```
Comment (Tisho99) [inline sale_order_line_clone/security/sale_order_line_clone_security.xml:7]: This should have one more tabulation

**Review by Tisho99 — CHANGES_REQUESTED** [review]
Some suggestions

I also think the pre-commit commit should have another name. Something like [IMP] pre-commit auto fixes of all modules

**Comment by luis-ron** [comment]
> > I am not sure if it would be like that, maybe Alberto can contribute, but in the README you indicate:
> > Go to Settings > Users and Company > Users and edit users how can copy sale order lines. Check the "Can Clone Sale Order Line" option in Extra Rights. Go to Sales > Products > Products and into the sales tab you can find the "Can Clone Sale Order Line" option.
> > As an addition u indicate: "how can copy sale order lines", it should be "who can copy and sale order lines" i think.
> > Maybe, you could indicate that the module adds a group that allows to indicate who can make use of the function offered by the module.
> > I mean, in a small company it is normal to configure users individually, but imagine in a large company... this would be done by groups, not by individual users.
> > I am waiting for your comments.
> > Thank you,
> 
> Thank you for your reply. All the user permission options are actually groups. If we look for the name of the option in the Users and Companies > Groups section, the group will appear so we can edit it and add users more quickly. I will add this point to the README if you think it is appropriate.

Okay, when you have the changes notify me and i'll review again :)

File: sale_order_line_clone/models/product_template.py (hunk 1 above)
Comment (RogerSans) [reply sale_order_line_clone/models/product_template.py]: Changed

File: sale_order_line_clone/models/product_template.py (hunk 2 above)
Comment (RogerSans) [reply sale_order_line_clone/models/product_template.py]: Deleted this part

File: sale_order_line_clone/README.rst (hunk 3 above)
Comment (RogerSans) [reply sale_order_line_clone/README.rst]: Changed

File: sale_order_line_clone/__manifest__.py (hunk 4 above)
Comment (RogerSans) [reply sale_order_line_clone/__manifest__.py]: Changed

File: sale_order_line_clone/models/sale_order.py (hunk 5 above)
Comment (RogerSans) [reply sale_order_line_clone/models/sale_order.py]: Changed

File: sale_order_line_clone/readme/CONFIGURE.md (hunk 6 above)
Comment (RogerSans) [reply sale_order_line_clone/readme/CONFIGURE.md]: Changed

File: sale_order_line_clone/security/sale_order_line_clone_security.xml (hunk 7 above)
Comment (RogerSans) [reply sale_order_line_clone/security/sale_order_line_clone_security.xml:7]: Changed

File: sale_order_line_clone/README.rst [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+Installation
+============
+
+#. Just install.
+
+Configuration
+=============
+
+#. Go to *Settings > Users and Company > Users* and edit users that can
```
Comment (Tisho99) [inline sale_order_line_clone/README.rst:40]: This section still is badly displayed.

Please, read this document: https://www.sygel.es/web#id=264&action=1638&model=document.page&view_type=form&cids=1&menu_id=996

You also need to add an space between #. blocks

File: sale_order_line_clone/security/sale_order_line_clone_security.xml (hunk 7 above)
Comment (Tisho99) [reply sale_order_line_clone/security/sale_order_line_clone_security.xml:7]: Sorry, the tabulation was between lines 6 and 9 (included), not between 3 and 10

File: sale_order_line_clone/models/product_template.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        compute="_compute_service_sale_line",
+        store=True,
+        readonly=False,
+    )
+
+    @api.depends("detailed_type")
+    def _compute_service_sale_line(self):
+        for record in self:
+            record.can_clone_sale_line = record.detailed_type != "service"
+        return False
```
Comment (Tisho99) [inline sale_order_line_clone/models/product_template.py]: Even though this causes no error, in compute functions returns are not needed

Comment (RogerSans) [reply sale_order_line_clone/models/product_template.py]: Removed

File: sale_order_line_clone/README.rst (hunk 8 above)
Comment (RogerSans) [reply sale_order_line_clone/README.rst:40]: Solved

File: sale_order_line_clone/security/sale_order_line_clone_security.xml (hunk 7 above)
Comment (Tisho99) [reply sale_order_line_clone/security/sale_order_line_clone_security.xml:7]: @RogerSans Can you review this?

Comment (RogerSans) [reply sale_order_line_clone/security/sale_order_line_clone_security.xml:7]: Done. Sry

**Review by Tisho99 — APPROVED** [review]
LGTM

**Comment by RogerSans** [comment]
@manuel-florido

**Pull request merged**